### PR TITLE
*: support auto analyze partition table

### PR DESCRIPTION
**Pull request by lamxTyler**
### What problem does this PR solve? 

Let auto analyze work for partition table.

### What is changed and how it works?

Support analyze table for a specific partition, the syntax is `alter table t analyze partition a, b index c, d` or `analyze table t partition a, b index c, d`

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - None

Related changes

 - None

### CODE REVIEW

**Comment by lamxTyler** [comment]
PTAL @coocood @winoros

File: statistics/update.go [hunk 1]
```diff
@@ -707,31 +707,52 @@ func (h *Handle) HandleAutoAnalyze(is infoschema.InfoSchema) error {
...
-					sql := fmt.Sprintf("analyze table %s index `%s`", tblName, idx.Name.O)
-					log.Infof("[stats] auto analyze index `%s` for table %s now", idx.Name.O, tblName)
-					return errors.Trace(h.execAutoAnalyze(sql))
+				continue
+			}
+			for _, def := range pi.Definitions {
+				sql := fmt.Sprintf("alter table %s analyze partition `%s`", tblName, def.Name.O)
+				statsTbl := h.GetPartitionStats(tblInfo, def.ID)
+				analyzed, err := h.autoAnalyzeTable(tblInfo, statsTbl, start, end, autoAnalyzeRatio, sql)
+				if analyzed {
```
Comment (coocood) [inline statistics/update.go]: Should we analyze all partitions at once instead of only one partition?

Comment (lamxTyler) [reply statistics/update.go]: I think only one partition is better because the partition stats is independent.

**Comment by lamxTyler** [comment]
PTAL @zz-jason

File: planner/core/planbuilder.go [hunk 2]
```diff
@@ -622,22 +623,46 @@ func getColsInfo(tn *ast.TableName) (indicesInfo []*model.IndexInfo, colsInfo []
...
+		found := false
+		for _, def := range pi.Definitions {
+			if def.Name.L == name.L {
+				found = true
+				ids = append(ids, def.ID)
+				break
+			}
+		}
+		if !found {
+			return nil, errors.New(fmt.Sprintf("Error in list of partitions to %s", tblInfo.Name.O))
```
Comment (zz-jason) [inline planner/core/planbuilder.go]: how about: "can not found the specified partition name xxx in the table definition"?

File: statistics/update.go [hunk 3]
```diff
@@ -707,31 +707,52 @@ func (h *Handle) HandleAutoAnalyze(is infoschema.InfoSchema) error {
...
+					return err
```
Comment (zz-jason) [inline statistics/update.go]: if one of the partition is analyzed, the rest of the partitions can never get a change to be analyzed, I think we should continue to analyze other partitions instead just return and terminate the analyze command.

File: statistics/update.go [hunk 4]
```diff
@@ -707,31 +707,52 @@ func (h *Handle) HandleAutoAnalyze(is infoschema.InfoSchema) error {
...
 				}
-				if _, ok := statsTbl.Indices[idx.ID]; !ok {
-					sql := fmt.Sprintf("analyze table %s index `%s`", tblName, idx.Name.O)
-					log.Infof("[stats] auto analyze index `%s` for table %s now", idx.Name.O, tblName)
-					return errors.Trace(h.execAutoAnalyze(sql))
+				continue
+			}
+			for _, def := range pi.Definitions {
+				sql := fmt.Sprintf("alter table %s analyze partition `%s`", tblName, def.Name.O)
```
Comment (zz-jason) [inline statistics/update.go]: It's wired that mysql uses the `alter table` statement to analyze the table partitions... https://dev.mysql.com/doc/refman/5.7/en/partitioning-maintenance.htm

despite the compatible issue, an we also support another syntax to analyze table partitions?

File: statistics/update.go (hunk 3 above)
Comment (lamxTyler) [reply statistics/update.go]: The reason that we only trigger one analyze once a time is that we can get the most updated auto analyze parameters. The rest of the partition can wait for the next round which is just 3s after.

File: statistics/update.go (hunk 4 above)
Comment (lamxTyler) [reply statistics/update.go]: Yes, we can also support syntax like `analyze table t partition`.

File: statistics/update.go (hunk 3 above)
Comment (eurekaka) [reply statistics/update.go]: > we can get the most updated auto analyze parameters.

What does this mean?

**Comment by eurekaka** [comment]
Conflicts detected with this refactor commit: https://github.com/pingcap/tidb/pull/7879

File: statistics/update.go (hunk 3 above)
Comment (lamxTyler) [reply statistics/update.go]: In https://github.com/pingcap/tidb/blob/master/statistics/update.go#L654, we get the parameters like analyze time period, so if we continue analyze other tables, we may not use the latest parameters.

**Pull request merged**